gdnative-core/src/user_data.rs

```diff
@@ -0,0 +1,386 @@
+//! Customizable user-data wrappers.
+//! 
+//! ## `NativeClass` and user-data
+//! 
+//! In Godot Engine, scripted behavior is attached to base objects through "script instances":
+//! objects that store script state, and allow dynamic dispatch of overridden methods. GDNative
+//! exposes this to native languages as a `void *` pointer, known as "user-data", that may point
+//! to anything defined by the native library in question.
+//! 
+//! Godot is written in C++, and unlike Rust, it doesn't have the same strict reference aliasing
+//! constraints. This user-data pointer can be aliased mutably, and called freely from different
+//! threads by the engine or other scripts. Thus, to maintain safety, wrapper types are be used
+//! to make sure that the Rust rules for references are always held for the `self` argument, and
+//! no UB can occur because we freed `owner` or put another script on it.
+//! 
+//! ## Which wrapper to use?
```

Comment:
This is a really nice summary, nice job!

We should probably add single-threaded versions of this later on, that way you don't have any overhead due to locks or atomics when using a `NO_THREADS` build of Godot or know in advance that no access via threads will be done.

Reply:
I think that is a good idea, combined with the default wrapper type alias. Though wrapper types that are not thread safe should be put behind a `no_threads` feature flag, imo, as it isn't `unsafe` to just slap an unsafe wrapper type on something that could be accessed by threads.